File: python/dataproc_templates/mongo/mongo_to_bq.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Sequence, Optional, Any
from logging import Logger
import argparse
import pprint

from pyspark.sql import SparkSession, DataFrameWriter

from dataproc_templates import BaseTemplate
from dataproc_templates.util.argument_parsing import add_spark_options
from dataproc_templates.util.dataframe_writer_wrappers import persist_dataframe_to_cloud_storage
import dataproc_templates.util.template_constants as constants


__all__ = ['MongoToBigQueryTemplate']


class MongoToBigQueryTemplate(BaseTemplate):
    """
    Dataproc template implementing exports from Mongo to BigQuery
    """

    @staticmethod
    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
        parser: argparse.ArgumentParser = argparse.ArgumentParser()

        parser.add_argument(
            f'--{constants.MONGO_BQ_INPUT_URI}',
            dest=constants.MONGO_BQ_INPUT_URI,
            required=True,
            help='Mongo Input Connection Uri'
        )
        parser.add_argument(
            f'--{constants.MONGO_BQ_INPUT_DATABASE}',
            dest=constants.MONGO_BQ_INPUT_DATABASE,
            required=True,
            help='Mongo Input Database Name'
        )
        parser.add_argument(
            f'--{constants.MONGO_BQ_INPUT_COLLECTION}',
            dest=constants.MONGO_BQ_INPUT_COLLECTION,
            required=True,
            help='Mongo Input Collection Name'
        )

        parser.add_argument(
            f'--{constants.MONGO_BQ_OUTPUT_DATASET}',
            dest=constants.MONGO_BQ_OUTPUT_DATASET,
            required=True,
            help='BigQuery Output Dataset Name'
        )

        parser.add_argument(
            f'--{constants.MONGO_BQ_OUTPUT_TABLE}',
            dest=constants.MONGO_BQ_OUTPUT_TABLE,
            required=True,
            help='BigQuery Output Table Name'
        )

        parser.add_argument(
            f'--{constants.MONGO_BQ_OUTPUT_MODE}',
            dest=constants.MONGO_BQ_OUTPUT_MODE,
            required=False,
            default=constants.OUTPUT_MODE_APPEND,
            help=(
                'BigQuery Output write mode '
                '(one of: append,overwrite,ignore,errorifexists) '
                '(Defaults to append)'
            ),
            choices=[
                constants.OUTPUT_MODE_OVERWRITE,
                constants.OUTPUT_MODE_APPEND,
                constants.OUTPUT_MODE_IGNORE,
                constants.OUTPUT_MODE_ERRORIFEXISTS
            ]

        )

        parser.add_argument(
            f'--{constants.MONGO_BQ_TEMP_BUCKET_NAME}',
            dest=constants.MONGO_BQ_TEMP_BUCKET_NAME,
            required=True,
            help='GCS Temp Bucket Name'
        )

        known_args: argparse.Namespace
        known_args, _ = parser.parse_known_args(args)

        return vars(known_args)

    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:

        logger: Logger = self.get_logger(spark=spark)

        # Arguments
        input_uri: str = args[constants.MONGO_BQ_INPUT_URI]
        input_database: str = args[constants.MONGO_BQ_INPUT_DATABASE]
        input_collection: str = args[constants.MONGO_BQ_INPUT_COLLECTION]
        output_mode: str = args[constants.MONGO_BQ_OUTPUT_MODE]
        big_query_output_dataset: str = args[constants.MONGO_BQ_OUTPUT_DATASET]
        big_query_output_table: str = args[constants.MONGO_BQ_OUTPUT_TABLE]
        big_query_temp_bucket: str = args[constants.MONGO_BQ_TEMP_BUCKET_NAME]

        ignore_keys = {constants.MONGO_BQ_INPUT_URI}
        filtered_args = {key: val for key,
                         val in args.items() if key not in ignore_keys}
        logger.info(
            "Starting Mongo to Big Query Spark job with parameters:\n"
            f"{pprint.pformat(filtered_args)}"
        )

        # Read
        input_data = spark.read \
            .format(constants.FORMAT_MONGO) \
            .option(constants.MONGO_INPUT_URI, input_uri) \
            .option(constants.MONGO_DATABASE, input_database) \
            .option(constants.MONGO_COLLECTION, input_collection) \
            .load()


        # Write
        input_data.write \
                .format(constants.FORMAT_BIGQUERY) \
                .option(constants.TABLE, big_query_output_dataset + "." + big_query_output_table) \
                .option(constants.TEMP_GCS_BUCKET, big_query_temp_bucket) \
                .mode(output_mode) \
                .save()

File: java/src/main/java/com/google/cloud/dataproc/templates/mongo/MongoToBQ.java
/*
 * Copyright (C) 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.mongo;

import static com.google.cloud.dataproc.templates.util.TemplateConstants.*;

import com.google.cloud.dataproc.templates.BaseTemplate;
import java.sql.SQLException;
import java.util.concurrent.TimeoutException;
import org.apache.commons.lang3.StringUtils;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.streaming.StreamingQueryException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MongoToBQ implements BaseTemplate {

  public static final Logger LOGGER = LoggerFactory.getLogger(MongoToBQ.class);

  private final String projectId;
  private final String inputURI;
  private final String inputDatabase;
  private final String inputCollection;
  private final String outputMode;
  private final String bqOutputDataset;
  private final String bqOutputTable;
  private final String bqTempBucket;

  public MongoToBQ() {

    projectId = getProperties().getProperty(PROJECT_ID_PROP);
    inputURI = getProperties().getProperty(MONGO_BQ_INPUT_URI);
    inputDatabase = getProperties().getProperty(MONGO_BQ_INPUT_DATABASE);
    inputCollection = getProperties().getProperty(MONGO_BQ_INPUT_COLLECTION);
    outputMode = getProperties().getProperty(MONGO_BQ_OUTPUT_MODE);
    bqOutputDataset = getProperties().getProperty(MONGO_BQ_OUTPUT_DATASET);
    bqOutputTable = getProperties().getProperty(MONGO_BQ_OUTPUT_TABLE);
    bqTempBucket = getProperties().getProperty(MONGO_BQ_TEMP_BUCKET_NAME);
  }

  @Override
  public void validateInput() throws IllegalArgumentException {

    if (StringUtils.isBlank(inputURI)
        || StringUtils.isBlank(inputDatabase)
        || StringUtils.isBlank(inputCollection)
        || StringUtils.isBlank(outputMode)
        || StringUtils.isBlank(bqOutputDataset)
        || StringUtils.isBlank(bqOutputTable)
        || StringUtils.isBlank(bqTempBucket)) {

      LOGGER.error(
          "{},{},{},{},{},{},{} is required parameter. ",
          MONGO_BQ_INPUT_URI,
          MONGO_BQ_INPUT_DATABASE,
          MONGO_BQ_INPUT_COLLECTION,
          MONGO_BQ_OUTPUT_MODE,
          MONGO_BQ_OUTPUT_DATASET,
          MONGO_BQ_OUTPUT_TABLE,
          MONGO_BQ_TEMP_BUCKET_NAME);
      throw new IllegalArgumentException(
          "Required parameters for MongoToBQ not passed. "
              + "Set mandatory parameter for MongoToBQ template "
              + "in resources/conf/template.properties file.");
    }
  }

  @Override
  public void runTemplate()
      throws StreamingQueryException, TimeoutException, SQLException, InterruptedException {

    // Initialize the Spark session
    SparkSession spark = SparkSession.builder().appName("Spark MongoToBQ Job").getOrCreate();

    // Read from Mongo
    Dataset<Row> rowDataset =
        spark
            .read()
            .format(MONGO_FORMAT)
            .option(MONGO_INPUT_URI, inputURI)
            .option(MONGO_DATABASE, inputDatabase)
            .option(MONGO_COLLECTION, inputCollection)
            .load();

    // Write to BigQuery
    rowDataset
        .write()
        .format(MONGO_BQ_FORMAT)
        .option(
            MONGO_BQ_TABLE, String.format("%s.%s.%s", projectId, bqOutputDataset, bqOutputTable))
        .option(MONGO_BQ_TEMP_BUCKET, bqTempBucket)
        .mode(outputMode.toLowerCase())
        .save();
  }
}
